src/ripple/shamap/impl/ShardFamily.cpp

```diff
@@ -0,0 +1,189 @@
+}
+
+std::shared_ptr<FullBelowCache>
+ShardFamily::getFullBelowCache(std::uint32_t ledgerSeq)
+{
+    auto const shardIndex{app_.getShardStore()->seqToShardIndex(ledgerSeq)};
+    std::lock_guard lock(fbCacheMutex_);
+    if (auto const it{fbCache_.find(shardIndex)}; it != fbCache_.end())
+        return it->second;
+
+    // Create a cache for the corresponding shard
+    auto fbCache{std::make_shared<FullBelowCache>(
+        "Shard family full below cache shard " + std::to_string(shardIndex),
+        stopwatch(),
+        cm_.collector(),
+        fullBelowTargetSize,
+        fullBelowExpiration)};
+    return fbCache_.emplace(shardIndex, std::move(fbCache)).first->second;
+}
+
+int
+ShardFamily::getFullBelowCacheSize()
+{
+    size_t sz{0};
+    std::lock_guard lock(fbCacheMutex_);
+    for (auto const& e : fbCache_)
+        sz += e.second->size();
+    return sz;
+}
+
+std::shared_ptr<TreeNodeCache>
+ShardFamily::getTreeNodeCache(std::uint32_t ledgerSeq)
+{
+    auto const shardIndex{app_.getShardStore()->seqToShardIndex(ledgerSeq)};
+    std::lock_guard lock(tnCacheMutex_);
+    if (auto const it{tnCache_.find(shardIndex)}; it != tnCache_.end())
+        return it->second;
+
+    // Create a cache for the corresponding shard
+    auto tnCache{std::make_shared<TreeNodeCache>(
+        "Shard family tree node cache shard " + std::to_string(shardIndex),
+        tnTargetSize_,
+        tnTargetAge_,
+        stopwatch(),
+        j_)};
+    return tnCache_.emplace(shardIndex, std::move(tnCache)).first->second;
+}
+
+std::pair<int, int>
+ShardFamily::getTreeNodeCacheSize()
+{
+    int cacheSz{0};
+    int trackSz{0};
+    std::lock_guard lock(tnCacheMutex_);
+    for (auto const& e : tnCache_)
+    {
+        cacheSz += e.second->getCacheSize();
+        trackSz += e.second->getTrackSize();
+    }
+    return {cacheSz, trackSz};
+}
+
+void
+ShardFamily::sweep()
+{
+    {
+        std::lock_guard lock(fbCacheMutex_);
+        for (auto it = fbCache_.cbegin(); it != fbCache_.cend();)
+        {
+            it->second->sweep();
+
+            // Remove cache if empty
+            if (it->second->size() == 0)
+                it = fbCache_.erase(it);
+            else
+                ++it;
+        }
+    }
+
+    std::lock_guard lock(tnCacheMutex_);
+    for (auto it = tnCache_.cbegin(); it != tnCache_.cend();)
+    {
+        it->second->sweep();
+
+        // Remove cache if empty
+        if (it->second->getTrackSize() == 0)
+            it = tnCache_.erase(it);
+        else
+            ++it;
+    }
+}
+
+void
+ShardFamily::reset()
+{
+    {
+        std::lock_guard lock(maxSeqMutex_);
+        maxSeq_ = 0;
+    }
+
+    {
+        std::lock_guard lock(fbCacheMutex_);
+        fbCache_.clear();
+    }
+
+    std::lock_guard lock(tnCacheMutex_);
+    tnCache_.clear();
+}
+
+void
+ShardFamily::missingNode(std::uint32_t seq)
+{
+    JLOG(j_.error()) << "Missing node in ledger sequence " << seq;
+
+    // Prevent recursive invocation
```

Comment:
I'm guessing you don't really mean "recursive" here.  If it were recursive then I think you'd need a `std:::recursive_mutex`, which you're not using.